Error handling for Observability code in browserstack service

sriteja777 opened the pull request:
## Proposed changes
Added error handling wrappers so as not impact test sessions in case of any errors in observability code

Related v7 PR: https://github.com/webdriverio/webdriverio/pull/10226

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [ ] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

sriteja777 commented:
@christian-bromann Please review and release this PR in your free time.

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -17,10 +18,18 @@ import type { UserConfig, UploadType, LaunchResponse, BrowserstackConfig } from
 import type { ITestCaseHookParameter } from './cucumber-types.js'
 import { BROWSER_DESCRIPTION, DATA_ENDPOINT, DATA_EVENT_ENDPOINT, DATA_SCREENSHOT_ENDPOINT } from './constants.js'
 import RequestQueueHandler from './request-handler.js'
+import * as util from 'node:util'
```
Let's move this to the top next to the other node imports
```suggestion
import util from 'node:util'
```

sriteja777 replied:
done https://github.com/webdriverio/webdriverio/pull/10183/commits/e8539f4a450860e198b32d961dc13d55fdfb7658

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -17,10 +18,18 @@ import type { UserConfig, UploadType, LaunchResponse, BrowserstackConfig } from
 const pGitconfig = promisify(gitconfig)
 const log = logger('@wdio/browserstack-service')
 
+/* User test config for build run minus PII */
+let userConfigForReporting: any = null
+let credentialsForCrashReportUpload: any = {}
```
Can these be typed?

amaanbs replied:
Done

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -97,7 +106,157 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
+            }
+        }
+    } catch (err) {
+        /* Wrong configuration like strings instead of json objects could break this method, needs no action */
```
can we log this?

amaanbs replied:
Will add.

chetan-p-browserstack replied:
done

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -97,7 +106,157 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
 }
 
-export async function launchTestSession (options: BrowserstackConfig & Options.Testrunner, config: Options.Testrunner, bsConfig: UserConfig) {
+function filterPII(userConfig: Options.Testrunner) {
```
What does PII mean? Can we have a comment what this function does?

amaanbs replied:
PII is user sensitive information present in config file which we filter out before sending config details. Will add the same as a comment.

chetan-p-browserstack replied:
added comment

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -97,7 +106,157 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
+    setCredentialsForCrashReportUpload(options, userConfig)
+}
+
+function processError(error: any, fn: Function, args: any) {
```
can we type this?

sriteja777 replied:
- `error`: The first argument is the `error` that gets passed from a catch block. Have checked the type we receive in catch block, that was of unknown type, so had to keep any as it is not a deterministic type
- `args`: These are the args that received to a function, and as this is a generic function and type of args not known before hand so had to keep any. Have moved this to `any[]` type, that will be more specific

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -97,7 +106,157 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
+    } catch (error) {
+        return log.error(`[Crash_Report_Upload] Failed to parse user credentials while reporting crash due to ${error}`)
+    }
+    if (!credentialsForCrashReportUpload.username || !credentialsForCrashReportUpload.password) {
```
We have the same `if` in line 216 so we could move this to the top of the function and save some code.

amaanbs replied:
The `if` in line 216 runs only once per worker to set the credentials global variable if username & password fields are not present on the current value of this variable or if this variable is an empty JSON as set initially. If fields username & password are not found in the parsed JSON object, we would like to return from this method so an additional check is needed.

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -97,7 +106,157 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
+    }
+}
+
+// https://tugayilik.medium.com/error-handling-via-try-catch-proxy-in-javascript-54116dbf783f
```
Can you explain why we need this?

sriteja777 replied:
This is a wrapper for error handling of methods in a class. Have [added comments](https://github.com/webdriverio/webdriverio/pull/10183/commits/e8539f4a450860e198b32d961dc13d55fdfb7658) in the code.

christian-bromann reviewed (commented):
Did an initial review

sriteja777 commented:
Hey @christian-bromann, me and @amaanbs addressed all review comments. Please re-review

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -97,7 +107,169 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
+        log.error(`[Crash_Report_Upload] Failed to parse user config while reporting crash due to ${error}`)
+    }
+
+    try {
+        const data = {
+            hashed_id: process.env.BS_TESTOPS_BUILD_HASHED_ID,
+            observability_version: {
+                frameworkName: 'WebdriverIO-' + (userConfigForReporting.framework || 'null'),
+                sdkVersion: bstackServiceVersion
+            },
+            exception: {
+                error: exception.toString(),
+                stackTrace: stackTrace
+            },
+            config: userConfigForReporting
+        }
+        const url = `${DATA_ENDPOINT}/api/v1/analytics`
+        const response: string = await got.post(url, {
+            ...DEFAULT_REQUEST_CONFIG,
+            ...credentialsForCrashReportUpload,
+            json: data
+        }).text()
+        log.debug(`[Crash_Report_Upload] Success response: ${JSON.stringify(response)}`)
+    } catch (error) {
+        log.error(`[Crash_Report_Upload] Failed due to ${error}`)
+    }
```
no need for try/catch:
```suggestion
    const data = {
        hashed_id: process.env.BS_TESTOPS_BUILD_HASHED_ID,
        observability_version: {
            frameworkName: 'WebdriverIO-' + (userConfigForReporting.framework || 'null'),
            sdkVersion: bstackServiceVersion
        },
        exception: {
            error: exception.toString(),
            stackTrace: stackTrace
        },
        config: userConfigForReporting
    }
    const url = `${DATA_ENDPOINT}/api/v1/analytics`
    return got.post(url, {
        ...DEFAULT_REQUEST_CONFIG,
        ...credentialsForCrashReportUpload,
        json: data
    }).text().then(
        (response: string) => log.debug(`[Crash_Report_Upload] Success response: ${JSON.stringify(response)}`),
        (error) => log.error(`[Crash_Report_Upload] Failed due to ${error}`)
    )
```

sriteja777 replied:
Done

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -13,14 +15,22 @@ import type { GitRepoInfo } from 'git-repo-info'
 const pGitconfig = promisify(gitconfig)
 const log = logger('@wdio/browserstack-service')
 
+/* User test config for build run minus PII */
+let userConfigForReporting: UserConfigforReporting = {}
+/* User credentials used for reporting crashes in browserstack service */
+let credentialsForCrashReportUpload: CredentialsForCrashReportUpload = {}
```
I am not sure how I feel about creating root variables to store this data. It's a code smell. This should be handled in a class, e.g. a class to handle crash reporting

sriteja777 replied:
removed, and moved to a separate class `crash-reporter`

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -97,7 +107,169 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
+function o11yErrorHandler(fn: Function) {
+    return function (...args: any) {
+        try {
+            // @ts-ignore
```
Why `@ts-ignore`?

sriteja777 replied:
Not needed, forgot to remove. Removed

sriteja777 commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -86,29 +93,34 @@ export default class BrowserstackService implements Services.ServiceInstance {
+                    this._isAppAutomate(),
+                    this._config.framework
+                )
+                await this._insightsHandler.before()
```
The constructor of insight handler can also raise an error.

The whole block of code was wrapped in try catch to ensure nothing breaks the tests from running even if there is unforeseen error in observability code. So even if, there is no chance of breaking now, but future code addition shouldn't break the user tests. So would like to wrap the entire code in try catch just to ensure that it doesn't break any user flow.

christian-bromann replied:
Nothing that breaks in the `before` hook will cause the test to break. You will have to throw a `SevereServiceError` to make the hook actual stop the test. The event listener calls can be moved outside of the `try/catch` because it is impossible that they throw.

sriteja777 commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -86,29 +93,34 @@ export default class BrowserstackService implements Services.ServiceInstance {
+                    'client:afterCommand',
+                    Object.assign(result, { sessionId: this._browser?.sessionId }),
+                    this._currentTest
+                ))
```
Done

sriteja777 commented:
Hey @christian-bromann, have addressed review comments. Please re-review.

Also tests are failing, but the failure doesn't seem related to this PR changes. They are failing on main branch itself.

chetan-p-browserstack commented:
@christian-bromann could you please prioritise review of this PR and another [PR](https://github.com/webdriverio/webdriverio/pull/10258) around hooks. Both needs to be rolled-out ASAP as we have commitments around it. If there are any review comments post your review, we will make sure to address it immediately.  Thanks in advance for your support.

sriteja777 commented:
Hey @christian-bromann sure will add the unit tests. 

and regarding the suggestion of moving browserstack service, yes it will help us in long term to keep it as a separate service. But there are some other things also that we need to evaluate for this. Will discuss this and get back to you. Thanks!

sriteja777 commented:
@christian-bromann Have added the unit tests. Please review and release

christian-bromann commented on packages/wdio-browserstack-service/src/crash-reporter.ts:
```diff
@@ -0,0 +1,121 @@
+
+const log = logger('@wdio/browserstack-service')
+const require = createRequire(import.meta.url)
+const { version: bstackServiceVersion } = require('../package.json')
```
This also happens in `launcher.js`, let's have this done once in one place e.g. a `constants.js` and export it for other packages to consume

sriteja777 replied:
done. moved to constants https://github.com/webdriverio/webdriverio/pull/10183/commits/fb0fbb92db57a39111250de63ee60149885b50aa

christian-bromann reviewed (commented):
Two last comments

sriteja777 commented:
@christian-bromann Have addressed the comments. Have a look!

christian-bromann reviewed (approved):
LGTM 👍

christian-bromann commented:
@sriteja777 please get back to me with the resolution your team made around moving the service to the BS organisation.

sriteja777 commented:
Yes @christian-bromann , currently we are exploring the solution you mentioned and other solutions that have no to minimal impact on our customers. We will get back to you asap with a solution for this .

cc @chetan-p-browserstack @sourav-kundu
